KAFKA-5473: handle ZK session expiration properly when a new session can't be established

prasincs opened the pull request:
(WIP: this commit isn't ready to be reviewed yet. I was checking the travis-ci build with the configuration changes in my account and opened the PR prematurely against trunk. I will make it consistent with Contribution guidelines once it's well tested.)

https://issues.apache.org/jira/browse/KAFKA-5473

Design:
`zookeeper.connection.retry.timeout.ms` => this determines how long to wait before triggering the shutdown. The default is 60000ms. 

Currently the implementation only handles the `handleSessionEstablishmentError` by waiting for the sessionTimeout.

prasincs committed:
simple change to KafkaHealth

dbolshak commented on core/src/main/scala/kafka/server/KafkaHealthcheck.scala:
```diff
@@ -117,14 +121,25 @@ class KafkaHealthcheck(brokerId: Int,
-      fatal("Could not establish session with zookeeper", error)
+      this.error("Session Establishment Error: %s, waiting for %d ms before exiting".format(error, connectionRetryMs))
+      if (reconnectingMs == -1){
+        reconnectingMs = Calendar.getInstance().getTimeInMillis()
```
why not `System.currentTimeMillis`?

prasincs replied:
After looking further I decided this design wont work to clean up the state. Cleaned it up further and using `System.currentTimeMillis` now.

prasincs committed:
Handle it for single case to validate the design

This is relying on the ZKClient to reconnect within the time set by
the config zookeeper.connection.retry.timeout.ms

If it doesn't get a state change within that time, it triggers a safe
shutdown for kafka. Previously it was just crashing

junrao commented:
@prasincs : Thanks for the patch. A couple of things. 

First, about the high level approach. Earlier, the proposal is to retry connecting for some time and then shut down the broker if the connecting fails. If we go with this approach, it seems that we could just reuse the existing zk.connection.time.ms instead of introducing a new one. An alternative approach is to just retry connecting forever until the ZK connection can be established w/o shutting down the broker. The benefit of the latter is that it can avoid expensive log recovery during server restart. So, it's worth considering this approach. In both approaches, it will be useful to have a metric that indicates the current state of ZK. Currently, we have a metric that reports the rate of various ZK events. However, it's hard to set an alert based on that. We will need a KIP to add the new ZK state metric.

Second, @onurkaraman has been working on a PR (https://github.com/apache/kafka/pull/3765) that will replace ZkUtils with a better ZK wrapper ZookeeperClient. We expect to merge the PR in the next couple of weeks. So, sequencing wise, it would be better for Onur's PR to get in first and then for your PR to use the new ZookeeperClient wrapper. This means that the patch will only be for trunk.

For the 1.0.0 branch, if needed, potentially we could just back port the new ZK state metric. That way, if the issue occurs, the admin at least be alerted.

Does that sound good to you?

prasincs commented:
Thanks @junrao, reusing `zk.connection.time.ms` sounds good to me. I think a more sustainable approach is what @onurkaraman is doing too. The sequence makes sense to me, get Onur's patch in and then I can update my patch accordingly to work with the new ZK state metric.

> potentially we could just back port the new ZK state metric. That way, if the issue occurs, the admin at least be alerted.

Are you suggesting to create a Gauge when session expiry event is received?

I don't think retrying forever by default is good idea, especially for cloud where you can have network partitions that manifest in unpredictable ways, between ZK and Kafka and between Kafka brokers. Ideally they're transient and go away but the default behavior should be to shut down quickly when clearly unhandle-able state is encountered and allow systems to alert/restart the service.

prasincs committed:
return if already waiting of ZK Reconnect

prasincs committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-5473

junrao commented:
@prasincs : As for monitoring, yes, I was thinking of a Gauge that tracks the current KeeperState. It seems that we already have such a gauge sessionStateGauge in kafka.server.KafkaHealthcheck. In the case when the new ZK session can't be established, that gauge will show DISCONNECTED. So, there is no need to add this to the 1.0.0 branch. We just need to maintain it properly using the new ZK wrapper.

As for retrying forever. I agree that network issue in the cloud is the common cause for ZK session expiration. When this happens, it's likely going to affect multiple and potentially all brokers. So, if we shut down the broker after the connection timeout, we could potentially bring down the whole cluster, which may not be ideal. On the other hand, if we retry forever, the cluster may still be able to work in the current state until the ZK issue is fixed. Note that bringing down a broker is not going to fix the ZK connectivity issue. So, the only benefit of bringing down the broker is just for notifying the admin. However, this can be achieved if the admin monitors the sessionStateGauge metric. The benefit of retrying forever is that (1) it allows the cluster to continue to operate when there is ZK session expiration, (2) it avoids the expensive log recovery due to a forced shutdown.

junrao commented:
On another thought, perhaps we could just introduce a new config zookeeper.reconnection.timeout.ms and wait for that amount and then shuts down the broker. An admin can set it to infinite if needed.

prasincs commented:
Yeah, `zookeeper.reconnection.timeout.ms` where 0 or -1 can mean infinite wait that the admin can wait if they choose to makes sense. The caller for ZK Client can just wait for that amount of time since a Session Expiry event until it receives a reconnect. 
I want to avoid the expensive log recovery if possible because of transient connection issues. If ZK cannot be connected to after the time out, cleanly shutting down is preferable to a rapid shutdown. Hopefully admins are monitoring sessionStateGauge and taking actions regardless.

junrao commented:
@prasincs : Since the new config requires a KIP, do you want to get the KIP discussion started? Thanks.

junrao commented:
@prasincs : Onur's PR (#3765) just got merged. Do you plan to resume the work on this PR soon? If so, there are a couple of things to be aware of.

1. Could you submit a KIP (for the new config? See https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Improvement+Proposals

2. The creation of a new ZK session is now moved to ZookeeperClient.initialize(). So, we need to pass in the new timeout config there. After a new session is created, it's probably better to complete the broker re-registration as the first thing in kafkaController.newSession() before it goes through controller election.

ijuma commented:
@junrao another thing to consider is that ZK doesn't re-resolve hosts when connection attempts fail:

https://issues.apache.org/jira/browse/ZOOKEEPER-2184

Looks like the new ZookeeperClient will recreate a ZooKeeper instance from scratch under certain cases, so maybe it works around the issue in ZOOKEEPER-2184, but we'd want to verify that.

junrao commented:
@ijuma : If a ZK client can't connect to ZK servers because it can't resolve hots, its session will be expired by the ZK server, but I am not sure if the client will be informed of the session expiration. So, the logic for handling session expiration probably won't cover that case.

What could happen is that if that client is the Kafka controller and it needs to access ZK, the operation will just block. The controller state will then be in a non-idle state. So, at least the admin can monitor and get alerted on such a state.

ijuma commented:
@junrao Do you know if DNS errors are propagated in some way? Because if they are, we could detect that and reinitialize the ZooKeeper instance. This is one of the most common issues in the cloud, so it would be good to have a solution.

junrao commented:
I am not sure, but my guess is that this is just propagated as a disconnect error.

junrao commented:
@prasincs : Do you think you could submit a KIP on the new config in the next couple of days? If not, I can submit one for you. Thanks.

prasincs committed:
Merge branch 'trunk' into KAFKA-5473

prasincs committed:
remove the timeout config for connection retry

and attempt to run the startup function if zk connection has been re-established

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 7959 tests run, 5 skipped, 1 failed.
--none--

prasincs commented:
@junrao Is the intent of KafkaZkClient to completely replace ZKUtils in 1.1.0? In that case I can implement the registerBrokerInZk that's in ZKUtils into KafkaZkClient  and use the same callback mechanism that @onurkaraman has to handle re-registering brokers/controllers on disconnects.

junrao commented:
@prasincs : Yes, the intent is to replace ZkUtils with KafkaZkClient. So, we want to add the re-creation logic to ZookeeperClient.initialize(). Once initialize() completes, ideally, we want to register the broker id before the controller does the re-election logic. So, one way to do this is to create a new ControllerEvent sth like RegisterBrokerAndReelect that first registers the broker id and then executes the logic in Reelect. In KafkaController, when calling zkClient.registerStateChangeHandler(), we enqueue the RegisterBrokerAndReelect event instead of the Reelect event in afterInitializingSession().

prasincs committed:
force register the broker Id until I re-implement the register brokerId

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

prasincs commented:
@junrao I don't see methods to create Ephemeral Znodes in the ZookeeperClient, and I don't think that re-registering BrokerId should be an async operation here. It's also likely I'm missing something here. I think the fastest to implement is to reuse ZKUtils and re-register the BrokerID in the initialize method in KafkaZKClient the way you have described?
This wont remove ZKUtils in 1.1.0 but should pave path for future? If @onurkaraman thinks otherwise, I can implement that too.

junrao commented:
@prasincs : Your idea sounds good. Note that even though we use the async ZK api in ZookeeperClient, ZookeeperClient.handleRequest() is actually blocking. So, we could implement the broker registration logic using ZookeeperClient.handleRequest() and call it after waitUntilConnected() in the construction of ZookeeperClient and after initialize().

asfgit commented:
FAILURE
 8067 tests run, 5 skipped, 2 failed.
--none--

prasincs committed:
Mocking out what I want to implement for re-registering broker in ZK

- I want to keep the separation of concerns between ZookeeperClient
  and KafkaZkClient
- But reconnects require some Kafka specific knowledge to re-register
  the brokerId before proceeding further
- This leads to somewhat ugly code in KafkaServer to keep references

prasincs commented:
@junrao that makes sense for implementing. The only problem I see is the current clean separation of concerns we have with KafkaZKClient and ZookeeperClient classes, where KafkaZkClient contains all the Kafka specific work. So, to keep the separation of concerns, I'm thinking of adding the reference to KafkaZKClient in ZookeeperClient that's only used in initialize time and implement all the Broker registration and Kafka specific work inside KafkaZKClient class. 
If this sounds reasonable, I'll move forward and implement the relevant parts in KafkaZKClient itself from ZKUtils.

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
SUCCESS 
 8081 tests run, 5 skipped, 0 failed.
--none--

onurkaraman commented:
@prasincs it seems to me that adding a reference to KafkaZkClient in ZooKeeperClient means you haven't separated concerns.

prasincs commented:
@onurkaraman yeah, it's still murky, that's why I just mocked the methods. Best way is to do everything Kafka related, including reconnects in KafkaZkClient. So could run another thread in KafkaZkClient that's subscriber of the ZK state changes and handle the reconnects? That'd likely be cleaner.

ijuma commented:
@prasincs can you not use a modified `StateChangeHandler`? Generally, you want to rely on some form of callback instead of circular references.

prasincs commented:
@ijuma yeah, that's what I'm thinking now too, references kind of solved my immediate problem but makes the code lot more brittle for maintenance. The current code does the broker id registration in KafkaHealth, leaning towards implementing it as callback there itself. So it can take KafkaZkClient instance and register the `StateChangeHandler` to bring it back to healthy state.

junrao commented:
@prasincs : Thanks for following up on this. There are a couple of issues with registering a separate StateChangeHandler. (1) There is no ordering guarantee. Currently, multiple state change handlers are executed in an arbitrary order. However, we want the broker registration to happen before the controller re-election process. (2) In general, we can't make any ZK call directly in the handler. This is because the handler is called inside ZooKeeperClientWatcher by the ZK event thread. If we make a ZK call there and get a ConnectionLossException, we need to call waitUntilConnected() and wait on isConnectedOrExpiredCondition. This condition is only notified inside ZooKeeperClientWatcher by the ZK event thread. So, we will deadlock in the case.

Given the above, it seems that the simplest approach is still what I recommended initially. Create a new ControllerEvent sth like RegisterBrokerAndReelect that first registers the broker id and then executes the logic in Reelect. In KafkaController, when calling zkClient.registerStateChangeHandler(), we enqueue the RegisterBrokerAndReelect event instead of the Reelect event in afterInitializingSession(). This means that the broker registration will be asynchronous. However, this is fine since this only thing we want to guarantee is that the controller election will only happen after the broker id is registered in ZK.

prasincs committed:
Revert "Mocking out what I want to implement for re-registering broker in ZK"

This reverts commit 0f4ebd888423e4513eadf07f2b916a7926c7a726.

junrao commented:
@prasincs : Do you still plan to work on this PR any time soon? This is the last one needed before we can completely switch to the new ZK wrapper on the server side. Thanks.

prasincs commented:
@junrao I'm somewhat unfamiliar with the interaction of ControllerEvent and haven't gotten chance to look at it in detail. I tried a few things locally but haven't tested to update the PR. If someone can do this quicker than I can, feel free to assign to them.

prasincs committed:
Merge branch 'trunk' into KAFKA-5473

prasincs committed:
implement BrokerId registration in KafkaZkClient and use that in KafkaHealthcheck

prasincs committed:
taking a stab at RegisterBrokerAndReelect ControllerEvent with KafkaZkClient reRegisterBrokerInZK()

prasincs commented:
@junrao I'm taking a stab at it this and next week, but I'm not sure if I can put a lot of time into it after next week. I added BrokerId registration using KafkaZKClient in the latest push. I'm not sure if the approach for what I'm doing in KafkaController is good yet. 
I have commented out the parts in KafkaHealthCheck pertaining to ZKClient state because I'm not certain if that code is superseded by the changes in #4265 cover it already or not.

junrao commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1398,6 +1398,21 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
     }
   }
 
+  case object RegisterBrokerAndReelect extends ControllerEvent {
```
We have the following code in KafkaController. We just need to replace Reelect with RegisterBrokerAndReelect in afterInitializingSession().

```
    zkClient.registerStateChangeHandler(new StateChangeHandler {
      override val name: String = StateChangeHandlers.ControllerHandler
      override def onReconnectionTimeout(): Unit = error("Reconnection timeout.")
      override def afterInitializingSession(): Unit = {
        eventManager.put(Reelect)
      }
```

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -587,6 +590,60 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
   }
 
+  // ZKStateChecker is started
+  class ZKStateChecker extends Thread {
```
This block of code is no longer needed.

junrao commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -57,6 +58,37 @@ class KafkaZkClient(zooKeeperClient: ZooKeeperClient, isSecure: Boolean) extends
+    checkedEphemeralCreate(brokerIdPath, brokerInfo)
+
+    info("Registered broker %d at path %s with addresses: %s".format(id, brokerIdPath, advertisedEndpoints.mkString(",")))
+  }
```
We probably can decouple the creation of brokerInfo from the registration in ZK. Perhaps we can add a method in ZkData.scala to create brokerInfo and have a method registerBrokerInZk(brokerInfo) to take in brokerInfo. In KafkaServer.startup(), we first create brokerInfo and call registerBrokerInZk(brokerInfo) to do the initial broker registration. We then pass in brokerInfo to KafkaController so that it can be used to re-register the broker when the ZK session expires.

prasincs replied:
Makes sense, passing BrokerInfo explicitly to Controller sounds better than keeping private variables in KafkaZkClient to recreate (that can be called by anyone and not just Controller)

junrao commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -39,12 +38,32 @@ import org.apache.zookeeper.AsyncCallback.{DataCallback, StringCallback}
+  val ConnectedReadOnly = Value("ConnectedReadOnly")
+  val SaslAuthenticated = Value("SaslAuthenticated")
+  val Expired = Value("Expired")
+}
```
We probably want to track these metrics in ZookeeperClient since that's where the ZK states are propagated. However, if you don't have the time, you can just leave this part out and we can address this in a separate jira.

prasincs replied:
Separate Jira sounds better to me since we already have one for the ZKClient metrics

junrao reviewed (commented):
@prasincs : Thanks for putting efforts on this. We are very close. A few comments below.

Also, not sure what's the best way to simulate ZK session expiration in a unit test. Perhaps you can at least test this manually by pausing and unpausing the broker (e.g. using kill -STOP and kill -CONT) after the session timeout.

prasincs commented:
@junrao yeah the pause and resume idea sounds good -- I I'll try with doing that and testing session expires -- like have 30 second expiry for Zookeeper Session and wait 31 seconds to send continue for the broker.

ijuma commented:
@prasincs, when do you think you'll be able to address the feedback? We're eager to complete the conversion to the new async zk client so that there's enough time to catch issues before the 1.1.0 release next month. Thanks!

prasincs commented:
@ijuma I'm aiming for next week. I saw the ZK metrics have been merged so the changes here should be relatively small now.

ijuma commented:
The metrics change that was merged is just for the latency metric. The session state is best done after this PR is merged (in a separate PR).

prasincs committed:
Merge branch 'trunk' into KAFKA-5473

prasincs committed:
WIP updates based on feedback on previous commits

This doesn't report the Session state yet -- that will be re-implemented soon.
- The code from KafkaHealthcheck was moved to KafkaServer to reuse the BrokerInfo to pass to Controller at startup time so that when the session expires, Conrtoller can re-register the broker and reconnect.

prasincs commented:
@ijuma @junrao I have updated with the suggestions. KafkaHealthCheck seemed unnecessary unless I put the session state there (tests are disabled too for now) but I didn't delete the file. BrokerInfo is created in KafkaServer and passed to KafkaController to Reelect. 
I tested with two brokers and single zk instance locally with `kill -STOP <zk pid>` and `kill -CONT <zk pid>` Do we have a place for tests like this?

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -292,12 +297,30 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
+        val jmxPort = System.getProperty("com.sun.management.jmxremote.port", "-1").toInt
+        val brokerInfo = new BrokerInfo(config.brokerId,
+          plaintextEndpoint.host, plaintextEndpoint.port,
+          listeners, jmxPort, config.rack, config.interBrokerProtocolVersion)
```
listeners should be updatedEndpoints.

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -97,6 +98,9 @@ object KafkaServer {
   private val isShuttingDown = new AtomicBoolean(false)
+  // this is set to true if the zookeeper client dies
+  private val isWaitingForZKReconnect = new AtomicBoolean(false)
+  private var zkState = ZKState.Unknown
```
These two variables are no longer used.

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -292,12 +297,30 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
 
         // Now that the broker id is successfully registered via KafkaHealthcheck, checkpoint it
         checkpointBrokerId(config.brokerId)
+        kafkaController.setBrokerInfo(brokerInfo)
```
We could just pass brokerInfo through the constructor of KafkaController and move the construction and the starting up of the controller to here.

prasincs replied:
I'm moving the code as is for listeners up and changing the KafkaController's constructor. This means the ports will be bound earlier (before controller starts). I'm not sure if that'll introduce race condition, but advertising the listeners in ZK before starting up could be worse.

junrao commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1411,6 +1417,21 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
+  case object RegisterBrokerAndReelect extends ControllerEvent {
+    override def state: ControllerState = ControllerState.ControllerChange
+
+    override def process(): Unit = {
```
Instead of duplicating the code in Reelect.process() here, we could probably just make RegisterBrokerAndReelect a subclass of Relect and override process() by calling super.process().

prasincs replied:
`Reelect` is a case object rather than case class, so can't subclass from case object.

junrao commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -73,6 +76,35 @@ class KafkaZkClient(zooKeeperClient: ZooKeeperClient, isSecure: Boolean, time: T
   }
 
+  private var brokerInfo: Array[Byte] = null;
+  private var brokerId: Int = -1;
```
Those variables are no longer needed.

prasincs replied:
removed

junrao commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -73,6 +76,35 @@ class KafkaZkClient(zooKeeperClient: ZooKeeperClient, isSecure: Boolean, time: T
+    info("Registered broker %d at path %s with addresses: %s".format(brokerInfo.ID(), brokerIdPath, brokerInfo.endpoints()))
+  }
+
+  def registerBrokerInZk(id: Int,
```
This method can now be removed.

prasincs replied:
Done

junrao commented on core/src/main/scala/kafka/server/KafkaHealthcheck.scala:
```diff
@@ -18,117 +18,49 @@
 
 import java.net.InetAddress
-import java.util.Locale
+import java.util.{Calendar, Locale}
```
We can just leave this class unchanged in the PR and clean it up in KAFKA-6320.

junrao commented on core/src/test/scala/unit/kafka/server/SessionExpireListenerTest.scala:
```diff
@@ -54,20 +54,20 @@ class SessionExpireListenerTest {
-    healthcheck.sessionExpireListener.handleStateChanged(Event.KeeperState.Disconnected)
-    checkMeterCount(expiresPerSecName, 1)
-    checkMeterCount(disconnectsPerSecName, 1)
+//    val healthcheck = new KafkaHealthcheck(brokerId, Seq.empty, zkUtils, None, ApiVersion.latestVersion, null)
```
We can just leave this class unchanged in the PR and clean it up in KAFKA-6320.

prasincs replied:
sounds good, I removed the registration steps from KafkaHealthCheck and tested. Rest is left as-is now

junrao commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -38,12 +38,32 @@ import org.apache.zookeeper.AsyncCallback.{DataCallback, StringCallback}
 
+// This is copy of the KeeperState with addition
+// of SessionEstablishmentError and NewSession
+object ZKState extends Enumeration{
```
We can just leave ZKState out in this PR and add it in ZkData in KAFKA-6320.

junrao commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -421,6 +421,7 @@ object KafkaConfig {
   val ZkSyncTimeMsDoc = "How far a ZK follower can be behind a ZK leader"
   val ZkEnableSecureAclsDoc = "Set client to use secure ACLs"
   val ZkMaxInFlightRequestsDoc = "The maximum number of unacknowledged requests the client will send to Zookeeper before blocking."
+
```
Changes in this file are not needed.

prasincs replied:
yeah, I'll replace with the file from trunk in next push

junrao commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -73,6 +76,35 @@ class KafkaZkClient(zooKeeperClient: ZooKeeperClient, isSecure: Boolean, time: T
+  private var brokerId: Int = -1;
+
+
+  def registerBrokerInZk(brokerInfo: BrokerInfo): Unit ={
```
space after =

prasincs replied:
fixed

junrao commented on core/src/main/scala/kafka/zk/ZkData.scala:
```diff
@@ -58,6 +58,30 @@ object BrokerIdsZNode {
   def encode: Array[Byte] = null
 }
 
+class BrokerInfo(id: Int,
```
To be consistent, perhaps this should be name BrokerIdZNode.

prasincs replied:
There's already a BrokerIdZNode in ZKData that BrokerInfo calls. It stores broker id, listeners and other info rather than being passed as args later to KafkaZkClient and consequently KafkaController.

junrao commented on core/src/main/scala/kafka/zk/ZkData.scala:
```diff
@@ -58,6 +58,30 @@ object BrokerIdsZNode {
+                 rack: Option[String],
+                 apiVersion: ApiVersion) {
+
+  def ID(): Int = {
```
Instead of defining this method. We can just make id val in the constructor.

prasincs replied:
done

junrao commented on core/src/main/scala/kafka/zk/ZkData.scala:
```diff
@@ -58,6 +58,30 @@ object BrokerIdsZNode {
+  def endpoints(): String = {
+    advertisedEndpoints.mkString(",")
+  }
+  def getBytes(): Array[Byte] = {
```
Add a new line above. Also, it's better to name the method encode().

prasincs replied:
done

junrao reviewed (commented):
@prasincs : Thanks for the updated patch. A few more comments below. As for the testing of re-registering the broker id, we can probably add a system test in the future.

prasincs committed:
remove all changes in KafkaConfig

prasincs committed:
keeping KafkaHealthCheck as is for this commit

prasincs committed:
rename getBytes method in ZKData.BrokerInfo to encode

prasincs committed:
use KafkaZkClient to register and renew -- remove broker register code in KafkaHealthCheck

prasincs committed:
zookeeper-server-stop.sh takes SIGNAL environment variable

so testing STOP and CONT is just about running

SIGNAL=STOP bin/zookeeper-server-stop.sh
SIGNAL=CONT bin/zookeeper-server-stop.sh

prasincs commented:
@junrao for testing, I've bin/zookeeper-server-stop.sh now takes a SIGNAL env variable that can be used to send the STOP and CONT signal, a simple system test could be to do something like this 
```
start zk
start kafka-server-0
start kafka-server-1
send STOP signal
wait 30 seconds
make sure neither of kafka servers die
send CONT signal
make sure brokers are registered and controller is elected
```

prasincs committed:
KafkaController takes the BrokerInfo in constructor

prasincs committed:
Merge branch 'trunk' into KAFKA-5473

junrao commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -39,12 +39,32 @@ import org.apache.zookeeper.AsyncCallback.{DataCallback, StringCallback}
+  val AuthFailed = Value("AuthFailed")
+  val ConnectedReadOnly = Value("ConnectedReadOnly")
+  val SaslAuthenticated = Value("SaslAuthenticated")
+  val Expired = Value("Expired")
```
We can revert the changes in this class and address them in KAFKA-6320.

prasincs replied:
Yep, reverted

junrao commented on core/src/main/scala/kafka/zk/ZkData.scala:
```diff
@@ -60,6 +60,26 @@ object BrokerIdsZNode {
+  def endpoints(): String = {
+    advertisedEndpoints.mkString(",")
+  }
+  def encode(): Array[Byte] = {
```
Add a new line above.

junrao commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -18,6 +18,8 @@ package kafka.zk
 import java.util.Properties
 
+import kafka.api.{ApiVersion, KAFKA_0_10_0_IV1, LeaderAndIsr}
+import kafka.cluster.{Broker, EndPoint}
```
unused imports Broker and LeaderAndIsr.

prasincs replied:
thanks for catching it. Removed

junrao commented on bin/zookeeper-server-stop.sh:
```diff
@@ -5,20 +5,20 @@
   exit 1
 else
-  kill -s TERM $PIDS
+  kill -s $SIGNAL $PIDS
```
Could we do that in bin/kafka-server-stop.sh too?

prasincs replied:
Added

junrao reviewed (commented):
@prasincs : Thanks for the updated patch. A few more minor comments below. Also, the following unit test failure reported by jenkins seems to be due to this patch.

kafka.server.ServerGenerateBrokerIdTest > testBrokerMetadataOnIdCollision FAILED
    org.scalatest.junit.JUnitTestFailedError: Expected exception java.lang.RuntimeException to be thrown, but org.apache.zookeeper.KeeperException$NodeExistsException was thrown
        at org.scalatest.junit.AssertionsForJUnit.newAssertionFailedException(AssertionsForJUnit.scala:100)
        at org.scalatest.junit.AssertionsForJUnit.newAssertionFailedException$(AssertionsForJUnit.scala:99)
        at org.scalatest.junit.JUnitSuite.newAssertionFailedException(JUnitSuite.scala:71)
        at org.scalatest.Assertions.intercept(Assertions.scala:812)
        at org.scalatest.Assertions.intercept$(Assertions.scala:802)
        at org.scalatest.junit.JUnitSuite.intercept(JUnitSuite.scala:71)
        at kafka.server.ServerGenerateBrokerIdTest.testBrokerMetadataOnIdCollision(ServerGenerateBrokerIdTest.scala:152)

        Caused by:
        org.apache.zookeeper.KeeperException$NodeExistsException: KeeperErrorCode = NodeExists
            at org.apache.zookeeper.KeeperException.create(KeeperException.java:119)
            at kafka.zk.KafkaZkClient.checkedEphemeralCreate(KafkaZkClient.scala:1304)
            at kafka.zk.KafkaZkClient.registerBrokerInZk(KafkaZkClient.scala:81)
            at kafka.server.KafkaServer.startup(KafkaServer.scala:278)
            at kafka.server.ServerGenerateBrokerIdTest.$anonfun$testBrokerMetadataOnIdCollision$1(ServerGenerateBrokerIdTest.scala:153)
            at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
            at org.scalatest.Assertions.intercept(Assertions.scala:805)
            ... 3 more

prasincs committed:
adding SIGNAL env variable for kafka-server-stop.sh too

prasincs committed:
revert ZkUtils to trunk

prasincs committed:
update brokerIdCollison test to catch NodeExistsException rather than base RuntimeException

prasincs commented:
@junrao thanks, I saw a lot of tests fail, I'll look at the remaining tests. For this particular one, since it was catching base RuntimeException, and that doesn't fit with new ZKClient's exceptions, I'm changing the test instead to be NodeExistsException. Since the intent is to check for broker metadata collision, I'm assuming that's fine because the alternate is to catch it and wrap as RuntimeException.

prasincs committed:
moving more of ZkClient errors with tests on BrokerId creation to NodeExists exception

junrao commented:
@prasincs : Thanks for the update. There seems to be the following test failure in the recent build.

```
kafka.server.ServerStartupTest > testConflictBrokerStartupWithSamePort FAILED
    kafka.common.KafkaException: Socket server failed to bind to localhost:35900: Address already in use.
        at kafka.network.Acceptor.openServerSocket(SocketServer.scala:348)
        at kafka.network.Acceptor.<init>(SocketServer.scala:273)
        at kafka.network.SocketServer$$anonfun$startup$1.apply(SocketServer.scala:99)
        at kafka.network.SocketServer$$anonfun$startup$1.apply(SocketServer.scala:91)
        at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at kafka.network.SocketServer.startup(SocketServer.scala:91)
        at kafka.server.KafkaServer.startup(KafkaServer.scala:241)
        at kafka.utils.TestUtils$.createServer(TestUtils.scala:135)
        at kafka.server.ServerStartupTest.testConflictBrokerStartupWithSamePort(ServerStartupTest.scala:63)

        Caused by:
        java.net.BindException: Address already in use
            at sun.nio.ch.Net.bind0(Native Method)
            at sun.nio.ch.Net.bind(Net.java:463)
            at sun.nio.ch.Net.bind(Net.java:455)
            at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
            at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
            at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:67)
            at kafka.network.Acceptor.openServerSocket(SocketServer.scala:344)
            ... 9 more
```

prasincs committed:
Catch the BindException and NodeExistsException explicitly in ServerStartupTest

junrao commented on core/src/test/scala/unit/kafka/server/ServerStartupTest.scala:
```diff
@@ -62,7 +65,7 @@ class ServerStartupTest extends ZooKeeperTestHarness {
       fail("Starting a broker with the same port should fail")
     } catch {
-      case _: RuntimeException => // expected
+      case _: BindException | NodeExistsException => // expected
```
BindException is the cause not the exception thrown. What's thrown is a KafkaException. So, you will want to catch that. Also, we can get rid of NodeExistsException.

prasincs replied:
Ah thanks for catching it, changing to KafkaException

prasincs committed:
change to KafkaException for conflicting broker with same port

junrao commented:
@prasincs : Thanks for the latest patch. LGTM

tedyu commented on core/src/main/scala/kafka/server/KafkaHealthcheck.scala:
```diff
@@ -117,18 +96,18 @@ class KafkaHealthcheck(brokerId: Int,
-      register()
-      info("done re-registering broker")
-      info("Subscribing to %s path to watch for new topics".format(ZkUtils.BrokerTopicsPath))
+      //info("re-registering broker info in ZK for broker " + brokerId)
```
I think commented out code should be removed.

junrao replied:
Yes, the plan is to do that in KAFKA-6320.
